src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/OnceInitializedOnceDisposedUnderLockAsync.cs

```diff
@@ -0,0 +1,94 @@
+        }
+
+        protected override sealed async Task DisposeCoreAsync(bool initialized)
+        {
+            await _semaphore.ExecuteAsync(() => DisposeCoreUnderLockAsync(initialized))
+                            .ConfigureAwait(true);
+
+            _semaphore.Dispose();
+        }
+
+        /// <summary>
+        ///     Disposes of managed and unmanaged resources owned by this instance, under a lock
+        ///     that prevents overlap with any currently executing actions passed to 
+        ///     <see cref="ExecuteUnderLockAsync(Func{CancellationToken, Task}, CancellationToken)"/>.
+        /// </summary>
+        /// <param name="initialized">
+        ///     A value indicating whether this instance had been previously initialized.
+        /// </param>
+        protected abstract Task DisposeCoreUnderLockAsync(bool initialized);
+
+        /// <summary>
+        ///     Executes the specified action under a lock that presents overlap with any currently executing actions passed to
+        ///     <see cref="ExecuteUnderLockAsync(Func{CancellationToken, Task}, CancellationToken)"/> and 
+        ///     <see cref="OnceInitializedOnceDisposedAsync.DisposeAsync"/>.
+        /// </summary>
+        /// <param name="action">
+        ///     The action to execute under the lock.
+        /// </param>
+        /// <param name="cancellationToken">
+        ///     The token to monitor for cancellation requests. The default value is <see cref="CancellationToken.None"/>.
+        /// </param>
+        /// <returns></returns>
+        /// <exception cref="OperationCanceledException">
+        ///     The result is awaited and <paramref name="cancellationToken"/> is cancelled.
+        ///     <para>
+        ///         -or-
+        ///     </para>
+        ///     The result is awaited and the s<see cref="ThreadSafeOnceInitializedOnceDisposedAsync"/> 
+        ///     has been disposed of.
+        /// </exception>
+        protected Task ExecuteUnderLockAsync(Func<CancellationToken, Task> action, CancellationToken cancellationToken = default)
+        {
+            Requires.NotNull(action, nameof(action));
+
+            return ExecuteUnderLockCoreAsync(action, cancellationToken);
+        }
+
+        private async Task ExecuteUnderLockCoreAsync(Func<CancellationToken, Task> action, CancellationToken cancellationToken = default)
+        {
+            using (var source = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, DisposalToken))
+            {
+                CancellationToken jointCancellationToken = source.Token;
+
+                try
+                {
+                    await _semaphore.ExecuteAsync(() => action(jointCancellationToken), jointCancellationToken)
+                                    .ConfigureAwait(true);
+                }
+                catch (ObjectDisposedException)
```

Comment:
This is speculative change based on code inspection - I can't figure out a way to actually repro this situation in a test, I don't have the right hooks.